Add qpax solver

lvjonok opened the pull request:
Hello Stéphane,

I have recently been exploring `jax`-based optimization and QP-solvers. During my exploration, I discovered [qpax](https://github.com/kevin-tracy/qpax) and noticed that it is not included in your repository. I would be glad to assist with benchmarking it against other solvers.

This PR is still in progress due to a pending PR in [qpax](https://github.com/kevin-tracy/qpax/pull/2). However, I believe the code here will not change significantly, and this work can proceed in parallel.

This PR also addresses the [need](https://github.com/qpsolvers/qpsolvers/issues/293) for a comparison of JAX-based solvers.

Thank you!

stephane-caron commented on qpsolvers/solvers/qpax_.py:
```diff
@@ -0,0 +1,68 @@
+
+    Paper: https://arxiv.org/pdf/2406.11749
+
+    # TODO: copy from paper
```
Would be great to add the paper to [doc/references.rst](https://github.com/qpsolvers/qpsolvers/blob/dc00ac33254e2dc2d8838d4fe2f473088a2034a3/doc/references.rst) and refer to it from here.

stephane-caron commented on qpsolvers/solvers/qpax_.py:
```diff
@@ -0,0 +1,68 @@
+from typing import Optional
```
```suggestion
#!/usr/bin/env python
# -*- coding: utf-8 -*-
#
# SPDX-License-Identifier: LGPL-3.0-or-later
# Copyright 2024 Lev Kozlov

"""Solver interface for `qpax <https://github.com/kevin-tracy/qpax>`__.

qpax is an open source QP solver that can be combined with JAX's jit and vmap
functionality, as well as differentiated with reverse-mode differentiation. It
is based on a primal-dual interior point algorithm. If you are using qpax in
some academic work, consider citing the corresponding paper [Tracy2024]_.
"""

from typing import Optional
```

Assuming the reference `[Tracy2024]` is added to doc/references.rst (see other comment below).

stephane-caron commented on tests/test_qpax.py:
```diff
@@ -0,0 +1,37 @@
+# -*- coding: utf-8 -*-
+#
+# SPDX-License-Identifier: LGPL-3.0-or-later
+# Copyright 2016-2024
```
```suggestion
# Copyright 2024 Lev Kozlov
```

The copyright is just a disclaimer (it applies whether we write it or not, so it's just written as an indication). If you write code as an individual your name goes there, and if you work for an institution your institute's name goes there (for instance I work for Inria so the code I write today is "Copyright 2024 Inria").

stephane-caron commented on qpsolvers/solvers/qpax_.py:
```diff
@@ -0,0 +1,68 @@
+    verbose: bool = False,
+    backend: Optional[str] = None,
+    **kwargs,
+) -> Solution:
```
We should put some documentation for this function. (It's very similar between solvers, don't hesitate to look at the others.)

stephane-caron reviewed (commented):
Some quick feedback here, nothing major.

stephane-caron commented:
Sounds great, thank you for looking into this and sharing with all users :smiley: 

I've taken a quick look at the present state of the PR and left some quick feedback points.

lvjonok commented:
Hi @stephane-caron,

I'm excited to share that qpax has been successfully released on PyPI! 🎉 I've also addressed the minor changes you pointed out in the PR. When you have a moment, could you please enable CI testing?

Thanks!

traversaro commented:
fyi @stephane-caron qpax is available on conda-forge as well (see https://github.com/conda-forge/staged-recipes/pull/27329 and https://github.com/conda-forge/qpax-feedstock).

stephane-caron commented on qpsolvers/solvers/__init__.py:
```diff
@@ -706,6 +706,48 @@
+    dense_solvers.append("qpax")
+    sparse_solvers.append("qpax")
+except ImportError as e:
+    raise e
```
From the PyPI CI workflow:

```

Run python -c "import qpsolvers"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/runner/work/qpsolvers/qpsolvers/qpsolvers/__init__.py", line 9, in <module>
    from ._internals import available_solvers
  File "/home/runner/work/qpsolvers/qpsolvers/qpsolvers/_internals.py", line 11, in <module>
    from .solvers import available_solvers as supported_solvers
  File "/home/runner/work/qpsolvers/qpsolvers/qpsolvers/solvers/__init__.py", line 749, in <module>
    raise e
  File "/home/runner/work/qpsolvers/qpsolvers/qpsolvers/solvers/__init__.py", line 742, in <module>
    from .qpax_ import qpax_solve_problem, qpax_solve_qp
  File "/home/runner/work/qpsolvers/qpsolvers/qpsolvers/solvers/qpax_.py", line 17, in <module>
    import qpax
ModuleNotFoundError: No module named 'qpax'
```

Here the change would be to simply pass, as we do with other not-found solvers:

```suggestion
    pass
```

stephane-caron reviewed (commented):
Thanks for the progress @lvjonok! You can now add `qpax` to `tox.ini` to run the qpax unit tests, and probably as well to `doc/environment.yaml` for the documentation workflow.

stephane-caron commented:
> fyi @stephane-caron qpax is available on conda-forge as well (see https://github.com/conda-forge/staged-recipes/pull/27329 and https://github.com/conda-forge/qpax-feedstock).

Thanks @traversaro. Let's take care of this afterwards in a follow-up PR to https://github.com/conda-forge/qpsolvers-feedstock.

lvjonok commented:
Hi @stephane-caron, I have updated the code according to your comments, waiting peacefully for the next review :)

stephane-caron commented:
Thank you for following up :+1: (There is still one unresolved comment regarding the docstring, but I can also take care of it when merging if you prefer.)

The main next step will be to make qpax pass all unit tests. Currently there is one failing API test:

```
ERROR: test_unbounded_below (tests.test_qpax.TestQpax.test_unbounded_below)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/runner/work/qpsolvers/qpsolvers/tests/test_qpax.py", line 34, in test_unbounded_below
    qpax_solve_qp(P, q)
TypeError: qpax_solve_qp() missing 6 required positional arguments: 'G', 'h', 'A', 'b', 'lb', and 'ub'
```

You can run tests locally from your repository by `tox -e py`. (Once your first contribution is merged the CI will also run by itself on your future PRs.)

Some other tests are just a matter of adjusting tolerances, for instance:

```
======================================================================
FAIL: test_warmstart_qpax (tests.test_solve_qp.TestSolveQP)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/scaron/tmp/qpsolvers/tests/test_solve_qp.py", line 675, in test
    self.assertLess(norm(x - known_solution), sol_tolerance)
AssertionError: 5.9604645e-08 not less than 1e-08
```

For this one you can just go to the test and update `sol_tolerance` to e.g. 1e-7 for qpax.

lvjonok commented:
Hello @stephane-caron, I have tuned to tolerances to pass the tests with `qpax`, however I have found that there are some problems which are feasible, but `qpax` raises an error.

I do not think it is a good sign for a solver to have such a behavior, I will look into the problems themselves more specifically, but it will further delay the PR.

stephane-caron commented on qpsolvers/solvers/qpax_.py:
```diff
@@ -29,26 +31,55 @@ def qpax_solve_problem(
+    n: int = q.shape[0]
+
+    if initvals is not None and verbose:
+        warnings.warn("warm-start values are ignored by PIQP")
```
```suggestion
        warnings.warn("warm-start values are ignored by qpax")
```

lvjonok commented:
Thanks, @stephane-caron, for guiding me!

I have carefully reviewed all the failing tests and reported problems in the original library.

Here is the list of issues:
- https://github.com/kevin-tracy/qpax/issues/4
- https://github.com/kevin-tracy/qpax/issues/5
- https://github.com/kevin-tracy/qpax/issues/6
- https://github.com/kevin-tracy/qpax/issues/7
- https://github.com/kevin-tracy/qpax/issues/8

To be honest, I am a bit disappointed that the solver failed so many times, but this is all part of the process of creating something new. Next, I will try to add some documentation to the core functions and look forward to the next review!

Thank you sincerely for your support!

lvjonok commented:
Hi @stephane-caron,

I believe I've addressed all the remaining points for this PR:

1. **Documentation:** I've added documentation to the core functions. While the quality might still need some refinement, the essential details are there.
2. **Linting:** The linting stage in GitHub Actions should now pass without issues.

I think everything is ready for a final review. Let me know if there's anything else I should look into.

Best,
Lev Kozlov

kevin-tracy commented:
Author of qpax here. These issues have been fixed in the latest release https://github.com/kevin-tracy/qpax/releases/tag/v0.0.7. However, there is one outstanding issue with https://github.com/kevin-tracy/qpax/issues/6 where the QP that fails is is nonconvex (has a negative eigenvalue in the quadratic cost matrix). 


On another note, since QPAX is written in JAX, it will default to using `float32` unless you directly specify `float64` with 
```python
import jax
jax.config.update("jax_enable_x64", True)
```
at the top of the file. While this isn't responsible for any of the failures noted above, it could potentially cause some issues in other tests down the line.

lvjonok commented:
I have enabled the tests corresponding to the solved issues. There are only two tests failing due mismatch in linear and box multipliers. Hope they will be resolved after https://github.com/kevin-tracy/qpax/issues/9

stephane-caron commented:
> However, there is one outstanding issue with https://github.com/kevin-tracy/qpax/issues/6 where the QP that fails is is nonconvex (has a negative eigenvalue in the quadratic cost matrix).

Thank you for taking a look :+1: I've checked the test and the eventual QP is convex, what happened here was a missing step in the transcription from LS to QP.

> On another note, since QPAX is written in JAX, it will default to using float32 unless you directly specify float64 with

Thanks for pointing this out. I'd recommend we should document this fact in the solver functions

lvjonok commented:
It seems that using `solution.is_optimal` is the best approach for future tests! 😊

During the update, I encountered a minor issue with version mismatches — it was necessary to bump the version in both `pyproject.toml` and `tox.ini`. After resolving this, everything is up and running. All `tox -e py` and `tox -e lint` checks pass successfully on my machine.

I did notice some tests failing on Windows. According to the official [JAX documentation](https://jax.readthedocs.io/en/latest/installation.html#install-cpu), support for Windows is still experimental. It might be worthwhile to consider removing `qpax` from the Windows pipelines.

Looking forward to your feedback and suggestions for the next steps!

stephane-caron commented:
LGTM! Thank you @lvjonok for following through all the way with this PR, making sure we properly interface with and test qpax :+1: And thank you @kevin-tracy for handling all issues and questions upstream.

I have merged the PR (disabling testing of qpax on Windows indeed). Next I will release qpsolvers v4.4.0 with the new solver interface.

lvjonok commented:
Thanks @stephane-caron!

I am really glad to contribute. Really proud of bringing something to the community, getting feedback and learning from others. Looking forward for new opportunities!)
